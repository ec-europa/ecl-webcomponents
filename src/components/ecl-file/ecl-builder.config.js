const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, '../../../node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  scripts: [
    {
      entry: path.resolve(nodeModules, '@ecl/vanilla-component-file/file.js'),
      dest: path.resolve(outputFolder, 'scripts/ecl-file-vanilla.js'),
      options: {
        banner,
        moduleName: 'ECL',
        includePaths,
        sourceMap: false,
      },
    },
  ],
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-file-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-file-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-file-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-file-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
      },
    },
  ],
};
