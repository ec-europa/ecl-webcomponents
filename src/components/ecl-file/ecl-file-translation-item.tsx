 import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'ecl-file-translations-item',
  shadow: false,
})

export class EclFileTranslationsItem{
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() variant: string = 'default';
  @Prop() meta: string;
  @Prop() fileTitle: string;
  @Prop() downloadLink: string;
  @Prop() downloadLabel: string;
  @Prop() language: string;

  getClass(): string {
    const styleClasses = [
      `ecl-file__translation-item`,
      `sc-ecl-file-${this.theme}`,
      this.styleClass
    ];

    return styleClasses.join(' ');
  }

  getChildrenClass() {
    return `ecl-file__translation-${this.variant === 'default' ? 'info' : 'detail'}`;
  }

  getTitle() {
    return <div 
            class={`ecl-file__translation-title sc-ecl-file-${this.theme}`}
            lang={this.language}
          >
            {this.fileTitle}
          </div>
  }

  getMeta() {
    return  <div class={`ecl-file__translation-meta sc-ecl-file-${this.theme}`}>
              {this.meta}
            </div>
  }

  render() { 
    return (
      <li
        class={this.getClass()}
      >
      { this.variant == 'thumbnail' ?
        <div class={`ecl-file__translation-detail sc-ecl-file-${this.theme}`}>
          {this.getTitle()}
        </div> : '' 
      }
      { this.variant == 'thumbnail' ?
        <div class={`ecl-file__translation-info sc-ecl-file-${this.theme}`}> 
          {this.getMeta()}
        </div> :
        <div class={`ecl-file__translation-info sc-ecl-file-${this.theme}`}>
          {this.getTitle()}
          {this.getMeta()}
        </div> 
      }
        <ecl-link
          path={this.downloadLink}
          variant="standalone"
          styleClass={`ecl-file__translation-download sc-ecl-file-${this.theme}`}
          theme={this.theme}
        >
          {this.downloadLabel}
          <ecl-icon 
            slot="icon-after"
            size="fluid"
            icon="download"
            theme={this.theme}
          >
          </ecl-icon>    
        </ecl-link>
      </li>
    );
  }
}
