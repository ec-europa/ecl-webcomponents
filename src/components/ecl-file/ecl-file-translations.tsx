import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'ecl-file-translations',
  shadow: false,
})

export class EclFileTranslations {
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() toggleLabel: string;
  @Prop() translationsTotal: number;
  @Prop() others: boolean = false;

  getClass(): string {
    const styleClasses = [
      `ecl-file__translation-container`,
      `sc-ecl-file-${this.theme}`,
      this.styleClass
    ];

    return styleClasses.join(' ');
  }

  render() { 
    return (
      <div 
        class={this.getClass()}
        data-ecl-file-translation-container
      >
        <ecl-button
          styleClass={`ecl-file__translation-toggle sc-ecl-file-${this.theme}`}
          variant="ghost"
          data-ecl-file-translation-toggle
          theme={this.theme}
        >
          {this.toggleLabel} { this.translationsTotal ? `(${this.translationsTotal})` : ''}
          <ecl-icon
            styleClass="ecl-button__icon ecl-button__icon--after"
            icon="corener-arrow"
            transform="rotate-180"
            size="fluid"
          >
          </ecl-icon>
        </ecl-button>
        <ul class={`ecl-file__translation-list sc-ecl-file-${this.theme}`}>
          <slot></slot>
        { this.others ?
          <li 
            class={`ecl-file__translation-item ecl-file__translation-description sc-ecl-file-${this.theme}`}
          >
            <slot name="others"></slot>
          </li> : '' }
        </ul>
      </div>
    );
  }
}
