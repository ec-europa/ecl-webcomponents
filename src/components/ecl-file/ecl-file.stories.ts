const getArgs = () => ({
  labels: false,
  variant: 'thumbnail',
  fileTitle: 'State of the Union 2018 brochure',
  downloadLabel: 'Download',
  downloadLink: '/example',
  language: 'English',
  meta: '(16.2 MB - PDF)',
  ariaLabel: 'Download file State of the Union 2018 brochure',
  description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer quis lorem
    tellus. Nullam sollicitudin suscipit diam, ac blandit ipsum tempor consectetur. Duis vitae pulvinar turpis.
    Donec maximus pharetra ex a ultricies.`,
});

const getArgTypes = () => ({
  labels: {
    type: { name: 'boolean' },
    description: 'Show file labels',
  },
  variant: {
    type: { name: 'select' },
    options: ['default', 'thumbnail'],
    description: 'File variants',
  },
  description: {
    type: { name: 'string' },
    description: 'only for the thumbnail variant'
  },
  fileTitle: {
    type: { name: 'string' },
    name: 'title of the file',
  },
  downloadLabel: {
    type: { name: 'string' },
    name: 'download link label',
    description: 'Label for the download button',
  },
  downloadLink: {
    type: { name: 'string' },
    name: 'download link',
    description: 'Link for the file download',
  },
  meta: {
    type: { name: 'string' },
    name: 'additional info',
    description: 'Meta information about the file',
  },
  language: {
    type: { name: 'string' },
    name: 'language of the file',
    description: 'Show the file language',
  },
  ariaLabel: {
    type: { name: 'string' },
    name: 'aria label of the element',
    description: 'To improve accessibility',
  },
});

export default {
  title: 'Components/file',
};


const TemplateWithTranslation = args =>
`<ecl-file
  variant="${args.variant}"
  theme="${args.theme}"
  file-title="${args.fileTitle}"
  download-link="${args.downloadLink}"
  download-label="${args.downloadLabel}"
  language="${args.language}"
  meta="${args.meta}"
  image="https://inno-ecl.s3.amazonaws.com/media/examples/example-image.jpg"
  image-alt="thumbnail alt"
  detail-meta='["Resource type", "Publication date"]'
  ecl-script
  taxonomies-label="Taxonomy list"
  taxonomies='["Taxonomy item 1",{"label": "Taxonomy item 2", "path": "/example"},"Taxonomy item 3"]'
  labels='${args.labels || args.variant === 'thumbnail' ? '[{"variant": "highlight", "label": "Highlight"}, {"variant": "medium", "label": "Medium"}]' : ''}'
>
  ${args.variant === 'thumbnail' ? args.description : ''}
  <ecl-file-translations
    theme="${args.theme}"
    toggle-label="Other languages"
    others
    slot="file-translations"
  >
    <ecl-file-translations-item
      language="bg"
      download-label="Download"
      download-link="/example"
      meta="(15.7 MB - PDF)"
      file-title="български"
      variant="${args.variant}"
      theme="${args.theme}"
    >
    </ecl-file-translations-item>
    <ecl-file-translations-item
      language="es"
      download-label="Download"
      download-link="/example"
      meta="(15.98 MB - PDF)"
      file-title="español"
      variant="${args.variant}"
      theme="${args.theme}"
    >
    </ecl-file-translations-item>
    <ecl-file-translations-item
      language="fr"
      download-label="Download"
      download-link="/example"
      meta="(15.98 MB - PDF)"
      file-title="français"
      variant="${args.variant}"
      theme="${args.theme}"
    >
    </ecl-file-translations-item>
    <p slot="others">
      Looking for another language which is not on the list? <a href="/example" class="ecl-link">Find out why</a>.
    </p>
  </ecl-file-translations>
</ecl-file>`;


export const FileWithTranslations = TemplateWithTranslation.bind({});
FileWithTranslations.storyName = 'File with translations';
FileWithTranslations.args = getArgs();
FileWithTranslations.argTypes = getArgTypes();
