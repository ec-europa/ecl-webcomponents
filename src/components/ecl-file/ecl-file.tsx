import { Component, Prop, h, Element, getAssetPath } from '@stencil/core';
declare const ECL: any;

@Component({
  tag: 'ecl-file',
  styleUrls: {
    ec: './build/styles/ecl-file-ec.css',
    eu: './build/styles/ecl-file-eu.css',
  },
  shadow: false,
  scoped: true,
})
export class EclFile {
  @Element() el: HTMLElement;
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() variant: string = 'default';
  @Prop() downloadLabel: string;
  @Prop() downloadLink: string;
  @Prop() fileTitle: string;
  @Prop() detailMeta: string;
  @Prop() language: string;
  @Prop() meta: string;
  @Prop() image: string;
  @Prop() imageAlt: string;
  @Prop() ariaLabel: string;
  @Prop() labels: string;
  @Prop() eclScript: boolean = false;

  getClass(): string {
    const styleClasses = [
      `ecl-file`,
      `ecl-file--${this.variant}`,
      this.styleClass
    ];

    return styleClasses.join(' ');
  }

  componentDidRender() {
    if (this.el.querySelector('.ecl-file__translation-list') && this.eclScript) {
      // Clean the html so that the script finds what it expects.
      const translations = this.el.querySelectorAll('.ecl-file__translation-item');
      this.el.querySelector('.ecl-file__translation-list').innerHTML = '';
      this.el.querySelector('.ecl-file__translation-list').append(...translations);
      // Load the ECL vanilla js if not already present.
      const src = getAssetPath('./build/scripts/ecl-file-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const file = new ECL.FileDownload(this.el.firstElementChild);
        file.init();
      };
      document.body.appendChild(script);
    }
  }

  getLabels() {
    const labelsArray = this.labels ? JSON.parse(this.labels) : '';
    let labelsMarkup = '';
    if (labelsArray) {
      labelsMarkup = labelsArray.map((label) => (
        <div class="ecl-file__label">
          <ecl-label 
            variant={label.variant}
            theme={this.theme}
          >
            {label.label}
          </ecl-label>
        </div>
      )); 
    }

    return labelsMarkup;
  }

  getMeta() {
    return <div class="ecl-file__meta">{this.meta}</div>
  }

  getDetailedMeta() {
    const detailMetaArray = this.detailMeta ? JSON.parse(this.detailMeta) : '';
    let detailMetaMarkup = '';

    if (detailMetaArray) {
      detailMetaMarkup = 
      <div class="ecl-file__detail-meta">
      {detailMetaArray.map((meta) => (
        <span class="ecl-file__detail-meta-item">{ meta }</span>
      ))} 
      </div>;
    }

    return detailMetaMarkup;
  }

  getTitle() {
    return <div class="ecl-file__title">{this.fileTitle}</div>
  }

  getLanguage() {
    return <div class="ecl-file__language">{this.language}</div> 
  }

  getDownload() {
    const downloadMarkup =
    <ecl-link
      path={this.downloadLink}
      variant="standalone"
    >
      {this.downloadLabel}
      <ecl-icon
        theme={this.theme}
        icon="download"
        size="fluid"
        slot="icon-after"
      ></ecl-icon>
    </ecl-link>;

    return downloadMarkup;
  }

  render() {
    const isDefault = this.variant == 'default';

    return (
      <div 
        class={this.getClass()}
        data-ecl-file
      >
        <div class="ecl-file__container">
        { isDefault ?
          <ecl-icon
            icon="file"
            size={this.theme === 'ec' ? '2xl' : 'm'}
            styleClass={`ecl-file__icon sc-ecl-file-${this.theme}`}
            theme={this.theme}
          ></ecl-icon> : 
          <div class="ecl-file__detail">
            <div class="ecl-file__detail-info">
              {this.getLabels()}
              {this.getDetailedMeta()}
              {this.getTitle()}
              <div class="ecl-file__description">
                <slot></slot>
              </div>
            </div>
          { this.image ?
            <img class="ecl-file__image" src={this.image} alt={this.imageAlt} /> : '' 
          }
          </div>
        }
        { isDefault ?
          <div class="ecl-file__info">
            {this.getLabels()}
            {this.getDetailedMeta()}
            {this.getTitle()}
            {this.getLanguage()}        
            {this.getMeta()}
          </div> :
          <div class="ecl-file__info">
            {this.getLanguage()}
            {this.getMeta()}
          </div>
        }
          {this.getDownload()}
        </div>
        <slot name="file-translations"></slot>
      </div>
    );
  }
}
