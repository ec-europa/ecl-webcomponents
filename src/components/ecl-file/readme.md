# ecl-file



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute        | Description | Type     | Default     |
| --------------- | ---------------- | ----------- | -------- | ----------- |
| `downloadLabel` | `download-label` |             | `string` | `undefined` |
| `downloadLink`  | `download-link`  |             | `string` | `undefined` |
| `fileTitle`     | `file-title`     |             | `string` | `undefined` |
| `language`      | `language`       |             | `string` | `undefined` |
| `meta`          | `meta`           |             | `string` | `undefined` |
| `styleClass`    | `style-class`    |             | `string` | `undefined` |
| `theme`         | `theme`          |             | `string` | `'ec'`      |
| `variant`       | `variant`        |             | `string` | `'default'` |


## Dependencies

### Depends on

- [ecl-link](../ecl-link)
- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-file-translations-item --> ecl-link
  ecl-file-translations-item --> ecl-icon
  ecl-link --> ecl-icon
  style ecl-file-translations-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
