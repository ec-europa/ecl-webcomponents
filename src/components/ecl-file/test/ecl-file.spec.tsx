import { newSpecPage } from '@stencil/core/testing';
import { EclFile } from '../ecl-file';
import { EclFileTranslations } from '../ecl-file-translations';
import { EclFileTranslationsItem } from  '../ecl-file-translation-item';

describe('ecl-file', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [EclFile, EclFileTranslations, EclFileTranslationsItem],
      html: `<ecl-file
        variant="default"
        theme="ec"
        file-title="State of the Union 2018 brochure"
        download-link="/example.html"
        download-label="Download"
        language="English"
        meta="(16.2 MB - PDF)"
        detail-meta='["Resource type", "Publication date"]'
        ecl-script
      >
        <ecl-file-translations
          theme="ec"
          translations-total="3"
          toggle-label="Other languages"
          others
        >
          <ecl-file-translations-item
            language="bg"
            download-label="Download"
            download-link="/example.html"
            meta="(15.7 MB - PDF)"
            fileTitle="български"
          >
          </ecl-file-translations-item>
          <ecl-file-translations-item
            language="es"
            download-label="Download"
            download-link="/example.html"
            meta="(15.98 MB - PDF)"
            fileTitle="español"
          >
          </ecl-file-translations-item>
          <ecl-file-translations-item
            language="fr"
            download-label="Download"
            download-link="/example.html"
            meta="(15.98 MB - PDF)"
            fileTitle="français"
          >
          </ecl-file-translations-item>
          <p slot="others">
            Looking for another language which is not on the list? <a href="/example.html" class="ecl-link">Find out why</a>.
          </p>
        </ecl-file-translations>
      </ecl-file>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
