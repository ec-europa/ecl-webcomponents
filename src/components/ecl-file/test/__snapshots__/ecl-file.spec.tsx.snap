// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-file renders 1`] = `
<ecl-file detail-meta="[&quot;Resource type&quot;, &quot;Publication date&quot;]" download-label="Download" download-link="/example.html" ecl-script="" file-title="State of the Union 2018 brochure" language="English" meta="(16.2 MB - PDF)" taxonomies="[&quot;Taxonomy item 1&quot;,{&quot;label&quot;: &quot;Taxonomy item 2&quot;, &quot;path&quot;: &quot;/example.html&quot;},&quot;Taxonomy item 3&quot;]" taxonomies-label="Taxonomy list" theme="ec" variant="default">
  <!---->
  <ecl-file-translations hidden="" others="" theme="ec" toggle-label="Other languages" translations-total="3">
    <!---->
    <div class="ecl-file__translation-container sc-ecl-file-ec" data-ecl-file-translation-container>
      <ecl-button data-ecl-file-translation-toggle styleclass="ecl-file__translation-toggle sc-ecl-file-ec" theme="ec" type="ghost">
        Other languages (3)
        <ecl-icon icon="corener-arrow" size="fluid" styleclass="ecl-button__icon ecl-button__icon--after" transform="rotate-180"></ecl-icon>
      </ecl-button>
      <ul class="ecl-file__translation-list sc-ecl-file-ec">
        <li class="ecl-file__translation-item sc-ecl-file-ec">
          <div class="ecl-file__translation-info sc-ecl-file-ec">
            <div class="ecl-file__translation-title sc-ecl-file-ec" lang="bg">
              български
            </div>
            <div class="ecl-file__translation-meta sc-ecl-file-ec">
              (15.7 MB - PDF)
            </div>
          </div>
          <ecl-link path="/example.html" styleclass="ecl-file__translation-download sc-ecl-file-ec" theme="ec" variant="standalone">
            Download
            <ecl-icon icon="download" size="fluid" slot="icon-after" theme="ec"></ecl-icon>
          </ecl-link>
        </li>
        <li class="ecl-file__translation-item sc-ecl-file-ec">
          <div class="ecl-file__translation-info sc-ecl-file-ec">
            <div class="ecl-file__translation-title sc-ecl-file-ec" lang="es">
              español
            </div>
            <div class="ecl-file__translation-meta sc-ecl-file-ec">
              (15.98 MB - PDF)
            </div>
          </div>
          <ecl-link path="/example.html" styleclass="ecl-file__translation-download sc-ecl-file-ec" theme="ec" variant="standalone">
            Download
            <ecl-icon icon="download" size="fluid" slot="icon-after" theme="ec"></ecl-icon>
          </ecl-link>
        </li>
        <li class="ecl-file__translation-item sc-ecl-file-ec">
          <div class="ecl-file__translation-info sc-ecl-file-ec">
            <div class="ecl-file__translation-title sc-ecl-file-ec" lang="fr">
              français
            </div>
            <div class="ecl-file__translation-meta sc-ecl-file-ec">
              (15.98 MB - PDF)
            </div>
          </div>
          <ecl-link path="/example.html" styleclass="ecl-file__translation-download sc-ecl-file-ec" theme="ec" variant="standalone">
            Download
            <ecl-icon icon="download" size="fluid" slot="icon-after" theme="ec"></ecl-icon>
          </ecl-link>
        </li>
        <li class="ecl-file__translation-description ecl-file__translation-item sc-ecl-file-ec">
          <p slot="others">
            Looking for another language which is not on the list?
            <a class="ecl-link" href="/example.html">
              Find out why
            </a>
            .
          </p>
        </li>
      </ul>
    </div>
  </ecl-file-translations>
  <div class="ecl-file ecl-file--default" data-ecl-file>
    <div class="ecl-file__container">
      <ecl-icon icon="file" size="2xl" styleclass="ecl-file__icon sc-ecl-file-ec" theme="ec"></ecl-icon>
      <div class="ecl-file__taxonomy">
        <ecl-description-list theme="ec" variant="horizontal">
          <ecl-description-list-term style-class="sc-ecl-file-ec" theme="ec">
            Taxonomy list
          </ecl-description-list-term>
          <ecl-description-list-definition items="[&quot;Taxonomy item 1&quot;,{&quot;label&quot;: &quot;Taxonomy item 2&quot;, &quot;path&quot;: &quot;/example.html&quot;},&quot;Taxonomy item 3&quot;]" style-class="sc-ecl-file-ec" theme="ec" type="taxonomy"></ecl-description-list-definition>
        </ecl-description-list>
      </div>
      <div class="ecl-file__info">
        <div class="ecl-file__detail-meta">
          <span class="ecl-file__detail-meta-item">
            Resource type
          </span>
          <span class="ecl-file__detail-meta-item">
            Publication date
          </span>
        </div>
        <div class="ecl-file__title">
          State of the Union 2018 brochure
        </div>
        <div class="ecl-file__language">
          English
        </div>
        <div class="ecl-file__meta">
          (16.2 MB - PDF)
        </div>
      </div>
      <ecl-link path="/example.html" variant="standalone">
        Download
        <ecl-icon icon="download" size="fluid" slot="icon-after" theme="ec"></ecl-icon>
      </ecl-link>
    </div>
  </div>
</ecl-file>
`;
